Add `searchAny` and `searchAll` functions

ahmadov opened the pull request:
### Changelog category (leave one):
- Experimental Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added functions `searchAny` and `searchAll` which are general purpose tools to search text indexes.

### Documentation entry for user-facing changes
- [ ] Documentation is written (mandatory for new features)

ahmadov committed:
Add new `searchAny` and `searchAll` functions

The new functions are repsonsible to search the needle tokens in the
tokenized text by the given tokenizer.

ahmadov committed:
Add `searchAny` and `searchAll` support for text index

These new functions can be used with the column defined in `text` index.

ahmadov committed:
Fix build

ahmadov committed:
fix search any condition for non-existent term

ahmadov committed:
add index analysis tests for search functions

ahmadov committed:
allow overwriting search function index parameters

ahmadov committed:
style fix and remove unused parameter

ahmadov committed:
Merge branch 'master' into ahmadov/search-function

rschu1ze reviewed (commented):
Just reviewed the first bits and posting comments. More comments to come.

ahmadov committed:
Move search function impl to cpp file

- Rename Search Props to Search Traits

ahmadov committed:
Merge remote-tracking branch 'origin/master' into ahmadov/search-function

ahmadov committed:
Fix after merging origin/master

ahmadov committed:
add search function tests for split tokenizer

ahmadov committed:
Add docs for `searchAny` and `searchAll` functions

ahmadov committed:
Fix CI build

rschu1ze reviewed (commented):
More comments and I am still not done reviewing the PR :-) It is just getting too late and I'll check the rest tomorrow.

rschu1ze commented on tests/queries/0_stateless/03403_function_search_index.sql:
```diff
@@ -0,0 +1,95 @@
+set allow_experimental_full_text_index = 1;
```
I see that some of the tests in CI are failing.

Example [here](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=80641&sha=latest&name_0=PR&name_1=Stateless+tests+%28amd_binary%2C+old+analyzer%2C+s3+storage%2C+DatabaseReplicated%2C+1%2F2%29&name_2=Tests) ... this is in a test that disables the analyzer. We shouldn't bother about that for new code. To fix this, you can force-enable the analyzer at the beginning of the test: `SET enable_analyzer = 1`.

Example [here](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=80641&sha=latest&name_0=PR&name_1=Stateless+tests+%28amd_binary%2C+ParallelReplicas%2C+s3+storage%29&name_2=Tests) This one tests with parallel replicas enabled which slightly alters the EXPLAIN output. To work around this, you can use `trim` in the same way as 02354_vector_search_pre_and_post_filtering.sql does.

rschu1ze commented on tests/queries/0_stateless/03403_function_search_index.sql:
```diff
@@ -0,0 +1,95 @@
+
+SELECT 'searchAny is used during index analysis';
+
+SELECT 'Skip index should choose none for non-existent term';
```
Thanks for explicitly spelling out the expected behavior.

rschu1ze commented on tests/queries/0_stateless/03403_function_search_index.sql:
```diff
@@ -0,0 +1,95 @@
+EXPLAIN indexes=1
+SELECT count() FROM tab WHERE searchAny(message, 'Click');
+
+SELECT 'Skip index should choose 1 part and 1024 granules out of 4 parts and 4096 granules';
```
`Skip` --> `Text`

rschu1ze commented on tests/queries/0_stateless/03403_function_search_index.sql:
```diff
@@ -0,0 +1,95 @@
+EXPLAIN indexes=1
+SELECT count() FROM tab WHERE searchAll(message, 'ClickHouse World');
+
+DROP TABLE tab;
```
As far as I understand, in above test, either all or none of the rows/granules are expected to match within each part. Maybe it makes sense to test the case that there are mixed matching/non-matching rows/granules within a part?

ahmadov committed:
Unify search function tests

ahmadov committed:
Update search impl

- Prefer to use `switch/case` instead of `if constexpr` for enums
- Styling changes
- Remove unnecessary parser and expression list entries

ahmadov commented on tests/queries/0_stateless/03403_function_search.sql:
```diff
@@ -0,0 +1,181 @@
+SELECT id from tab where searchAll(message, materialize('b')); -- { serverError ILLEGAL_COLUMN }
+-- search functions should be used with the index column
+SELECT id from tab where searchAny('a', 'b'); -- { serverError BAD_ARGUMENTS }
+-- SELECT id from tab where searchAny(col_str, 'b'); -- { serverError BAD_ARGUMENTS }
```
ah, good catch! I forgot to uncomment this line. The reason for it does not fail is this condition: https://github.com/ClickHouse/ClickHouse/blob/090e543b9afa75b23233462f6594d0853fe70355/src/Functions/search.cpp#L127-L128

since the table is empty, it does not check the next one which verifies the column is the index column. I fixed it by adding a single entry into table.

ahmadov replied:
This does not work unfortunately, because the function is called a few times with 0 row count (I guess during the analysis) and function parameters are passed only during the execution.

rschu1ze commented on src/Functions/search.cpp:
```diff
@@ -0,0 +1,196 @@
+                if (token == needles[pos])
+                    mask.add(pos);
+
+            if (mask.cardinality() == needles.size())
```
Yeah, let's limit to 64. I guess we can simply throw an exception in `executeImpl` of "searchAny/All".

ahmadov committed:
Limit number of supported needles to 64

- Update docs

ahmadov committed:
Add tests for mixed granules selection inside a part

ahmadov committed:
fix docs build

rschu1ze reviewed (approved):
Good stuff, thanks!

rschu1ze commented:
Hmm, somehow, there is still a result mismatch in [this CI run](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=80641&sha=latest&name_0=PR&name_1=Stateless+tests+%28amd_binary%2C+ParallelReplicas%2C+s3+storage%29&name_2=Tests) (parallel replicas again).

I am looking at the moment into https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=80641&sha=latest&name_0=PR&name_1=Build+%28fuzzers%29 which is a build error, probably some library link is missing.

ahmadov committed:
rename search.h/cpp to searchAnyAll.h/cpp

- disable parallel-replicas for the searchAny/searchAll functions for
  now.

ahmadov committed:
Fix fuzzers build

The `clickhouse_functions` dependency was missing in some fuzzer
executables. This is needed now because `MergeTreeIndexGin` now
depends on `searchAny` and `searchAll` functions.

ahmadov committed:
Extract text index search functions into its own library

To reduce the memory pressure on fuzzers, a new library
`clickhouse_functions_text` is created. Fuzzers depend on merge tree
index files and `MergeTreeIndexGin` depends on the `searchAny` and
`searchAll` functions.

Instead of including all functions (`clickhouse_functions`) in the
dependency, we can extract text index functions into its own library and
include only it in the fuzzers dependencies.
